server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java

```diff
@@ -1614,6 +1627,127 @@ public void testReservesBytesDuringPeerRecoveryPhaseOne() throws Exception {
+            }
+        }
+
+        try (var clusterStateSyncListeners = new ClusterStateSyncListeners()) {
+            final var primaryNodeTransportService = (MockTransportService) internalCluster().getInstance(
+                TransportService.class,
+                primaryNode
+            );
+            primaryNodeTransportService.addRequestHandlingBehavior(
+                Coordinator.COMMIT_STATE_ACTION_NAME,
+                (handler, request, channel, task) -> {
+                    assertThat(request, instanceOf(ApplyCommitRequest.class));
+                    clusterStateSyncListeners.getStateApplyDelayListener(((ApplyCommitRequest) request).getVersion())
+                        .addListener(
+                            ActionListener.wrap(ignored -> handler.messageReceived(request, channel, task), e -> fail(e, "unexpected"))
+                        );
+                }
+            );
+            primaryNodeTransportService.addRequestHandlingBehavior(
+                PeerRecoverySourceService.Actions.START_RECOVERY,
+                (handler, request, channel, task) -> {
+                    assertThat(request, instanceOf(StartRecoveryRequest.class));
+                    assertThat(((StartRecoveryRequest) request).clusterStateVersion(), greaterThan(initialClusterStateVersion));
+                    handler.messageReceived(
+                        request,
+                        new TestTransportChannel(
+                            new ChannelActionListener<>(channel).delegateResponse(
+                                (l, e) -> fail(e, "recovery should succeed on first attempt")
+                            )
+                        ),
+                        task
+                    );
+                    clusterStateSyncListeners.onStartRecovery();
+                }
+            );
+            clusterStateSyncListeners.addCleanup(primaryNodeTransportService::clearInboundRules);
+
+            final var replicaClusterService = internalCluster().getInstance(ClusterService.class, replicaNode);
+            final ClusterStateListener clusterStateListener = event -> {
+                final var primaryProceedListener = clusterStateSyncListeners.getStateApplyDelayListener(event.state().version());
+                final var indexRoutingTable = event.state().routingTable().index(indexName);
+                assertNotNull(indexRoutingTable);
+                final var indexShardRoutingTable = indexRoutingTable.shard(0);
+                if (indexShardRoutingTable.size() == 2 && indexShardRoutingTable.getAllInitializingShards().isEmpty() == false) {
+                    // this is the cluster state update which starts the recovery, so delay the primary node application until recovery
+                    // has started
+                    clusterStateSyncListeners.delayUntilRecoveryStart(primaryProceedListener);
+                } else {
+                    primaryProceedListener.onResponse(null);
```

Comment:
shouldn't this be called once the recovery starts through `clusterStateSyncListeners.onStartRecovery()`?

Reply:
This branch is for cluster state updates that _aren't_ the one which starts the update, particularly ones before the recovery, and we need to let the primary proceed with those updates straight away. I added a comment in e9ad47f77225c8a3acafeb7b34a078ffe2d2e589 which I hope helps.